src/hotspot/share/opto/node.hpp

```diff
@@ -1712,6 +1728,25 @@ class Unique_Node_List : public Node_List {
 #endif
 };
 
+// Definition must appear after complete type definition of Unique_Node_List
+template <typename Callback, typename Bypass>
+void Node::visit_uses(Callback callback, Bypass bypass) {
+  ResourceMark rm;
+  Unique_Node_List internals;
+  internals.push(this); // start traversal
+  for (uint j = 0; j < internals.size(); ++j) {
+    Node* internal = internals.at(j); // for every internal
+    for (DUIterator_Fast kmax, k = internal->fast_outs(kmax); k < kmax; k++) {
+      Node* internal_use = internal->fast_out(k);
+      if (bypass(internal_use)) {
+        internals.push(internal_use); // traverse this also
+      } else {
+        callback(internal_use);
```

Comment:
uses may now have the callback invoked multiple times, for example if we have something like this:

```
            this
             |
   +---------+--------+
   |                  |
   x                  y
   |                  |
   +--------+ +-------+
            | |
            use
```
This could not happen with Cast nodes (I think), but now that we generalize the function we should improve the logic, such that callback only gets called once.

Reply:
Yes, it is true that we sometimes may visit nodes multiple times. However, is it worth the extra bookkeeping overhead to ensure uses are visited only once? At least for the current applications of `visit_uses`, it's not unsound to (quite rarely, I think) add a node twice or more times to worklists.

I'm fine with either option, let me know what you think.

Reply:
I agree, it will be rare. But we don't know what it will be used for. One might want to count all uses or whatever, and then duplicates could matter.

You can also leave a clear comment, that uses may be called multiple times, then it is up to the user.

I would simply have renamed `internals` -> `worklist`.
And then push all nodes to it.
But when you look at a node, you first check it `is_boundary` -> `callback`.
Else put all its outputs on `worklist`.
I think the overhead is not that much.

Reply:
You are right, of course; we can use the `Unique_Node_List` to ensure we only visit uses once (I somehow managed to not read the `Unique` part of the class name and thought it was a basic stack). Then, the overhead is already there in any case and we can easily guarantee only one use. I'll make the change.